sdks/python/src/opik/integrations/dspy/callback.py

```diff
@@ -233,19 +236,51 @@ def on_lm_end(
             exception=exception,
             outputs=outputs,
         )
-        self._end_trace(call_id=call_id)
+
+    def on_tool_start(
+        self,
+        call_id: str,
+        instance: Any,
+        inputs: Dict[str, Any],
+    ) -> None:
+        span_data = self._collect_common_span_data(instance, inputs)
+        self._map_call_id_to_span_data[call_id] = span_data
+        self._set_current_context_data(span_data)
+
+    def on_tool_end(
+        self,
+        call_id: str,
+        outputs: Optional[Dict[str, Any]],
+        exception: Optional[Exception] = None,
+    ) -> None:
+        self._end_span(
+            call_id=call_id,
+            exception=exception,
+            outputs=outputs,
+        )
 
     def flush(self) -> None:
         """Sends pending Opik data to the backend"""
         self._opik_client.flush()
 
-    def _callback_context_set(self, value: ContextType) -> None:
-        token = self._current_callback_context.set(value)
-        self._map_span_id_or_trace_id_to_token[value.id] = token
+    def _set_current_context_data(self, value: ContextType) -> None:
+        if isinstance(value, span.SpanData):
+            self._context_storage.add_span_data(value)
+        elif isinstance(value, trace.TraceData):
+            self._context_storage.set_trace_data(value)
+        else:
+            raise ValueError(f"Invalid context type: {type(value)}")
+
+    def _get_current_context_data(self) -> Optional[ContextType]:
+        if span_data := self._context_storage.top_span_data():
+            return span_data
+        return self._context_storage.get_trace_data()
 
     def _get_span_type(self, instance: Any) -> types.SpanType:
         if isinstance(instance, dspy.Predict):
             return "llm"
         elif isinstance(instance, dspy.LM):
             return "llm"
+        elif isinstance(instance, dspy.Tool):
```

Comment:
Sometimes in can be dsp.ChainOfThought. I think that is the only one used in our optimizer implementation.